session.go

```diff
@@ -0,0 +1,160 @@
+type Config struct {
+	ClientID     string
+	ClientSecret string
+	Keypair      *Keypair
+	AccessTTL    int64
+	RefreshTTL   int64
+	Issuer       string
+}
+
+// NewSession returns a pointer to a new Session struct
+func (ss *SessionStore) NewSession(scope string, oAuthState string, oidcNonce string, user *User) (*Session, error) {
+	sessionID, err := nonce(24)
+	if err != nil {
+		return nil, err
+	}
+
+	session := &Session{
+		SessionID:  sessionID,
+		Scopes:     strings.Split(scope, " "),
+		OAuthState: oAuthState,
+		OIDCNonce:  oidcNonce,
+		User:       user,
+	}
+
+	ss.Store[sessionID] = session
+
+	return session, nil
+}
+
+// AccessToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) AccessToken(config *Config, timeFunc TimeNowFunc) (string, error) {
+
+	claims := standardClaims(s, config, config.AccessTTL, timeFunc)
+
+	token, err := config.Keypair.SignJWT(claims)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// standardClaims returns a populated jwt.StandardCLaims struct
+func standardClaims(s *Session, config *Config, ttl int64, timeFunc TimeNowFunc) jwt.StandardClaims {
+	if timeFunc == nil {
+		timeFunc = TimeFunc
+	}
+
+	return jwt.StandardClaims{
+		Audience:  "Audience",
+		ExpiresAt: timeFunc().UnixNano() + ttl,
+		Id:        s.SessionID,
+		IssuedAt:  timeFunc().UnixNano(),
+		Issuer:    config.Issuer,
+		NotBefore: timeFunc().UnixNano(),
+		Subject:   s.User.Email,
+	}
+}
+
+// RefreshToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) RefreshToken(config *Config, timeFunc TimeNowFunc) (string, error) {
+	claims := standardClaims(s, config, config.RefreshTTL, timeFunc)
+
+	token, err := config.Keypair.SignJWT(claims)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// IDTokenClaims are the claims to be provided to jwt.Sign to create the ID token
+type IDTokenClaims struct {
+	PreferredUsername string   `json:"preferred_username,omitempty"`
+	Email             string   `json:"email,omitempty"`
+	Phone             string   `json:"phone_number,omitempty"`
+	Address           string   `json:"address,omitempty"`
+	Groups            []string `json:"groups,omitempty"`
+	EmailVerified     *bool    `json:"email_verified,omitempty"`
+	jwt.StandardClaims
+}
+
+// IDToken returns the JWT token with the appropriate claims for an access token
+func (s *Session) IDToken(config *Config, timeFunc TimeNowFunc) (string, error) {
+	claims := standardClaims(s, config, config.AccessTTL, timeFunc)
+
+	var idClaims IDTokenClaims
+
+	idClaims.Audience = "Audience"
+	idClaims.ExpiresAt = claims.ExpiresAt
+	idClaims.Id = claims.Id
+	idClaims.IssuedAt = claims.IssuedAt
+	idClaims.Issuer = claims.Issuer
+	idClaims.NotBefore = claims.NotBefore
+	idClaims.Subject = claims.Subject
+
+	idClaims.PreferredUsername = s.User.PreferredUsername
+
+	if contains(s.Scopes, "address") {
```

Comment:
I think this might be cleaner if we loop through the scopes array and do the partial population based on the particular scope.

We'll also need to plan on doing this same logic later in the `userinfo` endpoint based on the Session we pull up from the SessionStore using the AccessToken we are given.

So this is a good candidate for a helper method since we will need the logic in both places.

Reply:
Other intriguing option: We have a method on the User that takes in the scopes and clones a version of itself with any data that isn't applicable for the scopes given. Then we are set for generic consumption of a User by this and `userinfo` 🤔